Update/remove duplicate post checkout redirect logic

niranjan-uma-shankar opened the pull request:
### Changes proposed in this Pull Request

* This PR refactors the concierge upsell related code and some parts of checkout to remove some code duplication.
* We have two versions of the Support Session page - `/checkout/SITE_SLUG/add-support-session` and `/checkout/SITE_SLUG/offer-quickstart-session` - that add the same product(concierge-session) to cart but differ in their copies. This PR consolidates the common parts of the code, so that it will be easier to add future copy variations without introducing code duplication or complexity.

Since there are a bunch of changes that have gone in, I'm summarizing them below:

### In [client/my-sites/checkout/index.js](https://github.com/Automattic/wp-calypso/pull/33996/files#diff-137f56ac9700496d02aebf4e8ad62b04)

1. Rename `add-support-session` slug to `offer-support-session`, since that route only shows the offer page and does not add anything to cart. This will make it in line with the `offer-quickstart-session` slug. 
2. Add the callback `conciergeSessionNudge()` for all the offer page routes. This callback consolidates the separate handlers we had for the Quick Start and Support Session versions. 

### In [client/my-sites/checkout/controller.jsx](https://github.com/Automattic/wp-calypso/pull/33996/files#diff-7f164a3b7deeca0666c8602e2b20e4ac):

1. Have just one callback `conciergeSessionNudge ()` handle both the Quick Start and Support Session offer routes.
2. Use the `<CheckoutContainer>` HOC for the concierge offer components.

### In [client/my-sites/checkout/checkout/checkout-container.jsx](https://github.com/Automattic/wp-calypso/pull/33996/files#diff-29f2e23b8d6785007185724253016b67):

1. This HOC will clear transaction data for those components that need it by calling the `<CartData>` HOC. 
2. The Order Summary a.k.a Cart will be shown by default, as determined by `shouldShowCart`. The concierge upsell offer pages will define this value as false.
3. Most importantly, it makes the concierge upsell nudges a child of the `<Checkout>` component, so that we can re-use the post checkout redirect handlers present in `<Checkout>`.

### In [client/my-sites/checkout/concierge-session-nudge/index.jsx](https://github.com/Automattic/wp-calypso/pull/33996/files#diff-cbd823e550be2fae3889881a67178394):

This component acts as a parent component for the concierge offer page. It has two children - `ConciergeQuickstartSession` and `ConciergeSupportSession`. Each of these two child components differ in the copy and design of the offer page. All button click handlers are placed in the parent.

### In [client/my-sites/checkout/concierge-session-nudge/style.scss](https://github.com/Automattic/wp-calypso/pull/33996/files#diff-3666f7629342640f15756c66be9cc715):

This file now only has the CSS required for the parent component. The content specific CSS is moved to the child components.

### In [client/my-sites/checkout/concierge-session-nudge/concierge-quickstart-session](https://github.com/Automattic/wp-calypso/pull/33996/files#diff-36ac488a973a8db6edccec87c8050802) AND [client/my-sites/checkout/concierge-session-nudge/concierge-support-session/ :

These are the child components of the concierge upsell nudge that use the Quick Start version / Support session version of the copy, as can be seen in the `header()`, `body()` and `footer()` functions. The CSS specific to this component is in the style.scss file in the same folder.

Both these component are implemented as Pure Components.

Given that each of these component is now a child of `<Checkout>` (thanks to <CheckoutContainer> HOC) we can re-use the post checkout handlers present in `Checkout` for Accept and Decline button actions.

### In [client/my-sites/checkout/checkout/index.jsx](https://github.com/Automattic/wp-calypso/pull/33996/files?file-filters%5B%5D=.js&file-filters%5B%5D=.jsx#diff-199e3967ae413c2ff48db966abbf8386):

We use the post checkout handler when the Accept of Decline buttons are clicked on the upsell nudges.

### In [client/state/selectors/is-eligible-for-checkout-to-checklist.js](https://github.com/Automattic/wp-calypso/pull/33996/files?file-filters%5B%5D=.js&file-filters%5B%5D=.jsx#diff-349f431f9f09abc95dc6816090be8958):

When the Decline button is clicked on any of the concierge offer nudges, the cart will be empty. So an additional check here ensures that even if the cart is empty, we still want to continue checking eligibility for redirecting to Checklist (if the user is coming via signup and purchasing a plan, then we wan to show the Checklist).

#### Why is GSuite upsell code not being updated?

Since that part of the code would require coordination with another team, I plan to implement the GSuite changes alone as a follow up PR after this one ships.

### Copy change

~When the upsell nudge is shown right after signup(i.e with a receipt ID in the URL), the header text is changed from `Hold tight, your site is being upgraded` .to `A special one time offer, just for you` - [#](https://github.com/Automattic/wp-calypso/pull/33996/files#diff-36ac488a973a8db6edccec87c8050802R60) and [#](https://github.com/Automattic/wp-calypso/pull/33996/files#diff-c726a2367c91b5e6b02938c834ad1821R55).~

### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Scenario 1**

* Signup for a new WPCOM account at /start; Put yourself in the `offer` variation of the `conciergeUpsellDial ` AB test.
* Purchase a plan(not business or eCommerce) and verify that the Quick Start session offer page is shown
* Verify that you are able to purchase the quick start session successfully

**Scenario 2**

* In any logged in account, open the quick start offer with a direct link - `http://calypso.localhost:3000/checkout/{SITE_SLUG}/offer-quickstart-session` and verify that you are able to purchase

* * In any logged in account, open the support session offer with a direct link - `http://calypso.localhost:3000/checkout/{SITE_SLUG}/offer-support-session` and verify that you are able to purchase

**Scenario 3**

* While logged out of WPCOM, open the quickstart offer page `http://calypso.localhost:3000/checkout/{SITE_SLUG}/offer-quickstart-session`. Clicking on the `Get Started` button, verify that you are able to complete purchase of the concierge session.

southp commented on client/my-sites/checkout/concierge-session-nudge/concierge-support-session/index.jsx:
```diff
@@ -0,0 +1,221 @@
+		const fullCost = Math.round( productCost * 1.5 );
+		const savings = fullCost - productCost;
+		return (
+			<Fragment>
```
Since we have introduced `<>` syntax above already, I'd suggest us to replace `<Fragment>` with that overall.

niranjan-uma-shankar replied:
Done in [a5da8b9](https://github.com/Automattic/wp-calypso/pull/33996/commits/a5da8b95b9c91a397e4e77aeaeb2c84c8451fbe5).

southp commented on client/state/selectors/is-eligible-for-checkout-to-checklist.js:
```diff
@@ -21,14 +26,16 @@ import isEligibleForDotcomChecklist from './is-eligible-for-dotcom-checklist';
-		hasEcommercePlan( cart )
-	) {
-		return false;
+	if ( ! isEmpty( getAllCartItems( cart ) ) ) {
```
Since the subsequent cart selectors, e.g. `hasDomainMapping()`, `hasDomainRegistration()`, etc., are all calling `getAllCartItems()` to check already, do we need this safety check?

niranjan-uma-shankar replied:
If the `Decline` button is clicked on the concierge nudge([here](https://github.com/Automattic/wp-calypso/pull/33996/files/7e54f5251bbf30c1333429ef72d14e6f730a5089#diff-cbd823e550be2fae3889881a67178394R145)), then not having the `isEmpty( getAllCartItems( cart )` check will make the function `isEligibleForCheckoutToChecklist ()` return false, since:
`! hasPlan( cart ) && ! hasConciergeSession( cart ) )`

will evaluate to true. We don't want it to return false, since then we would end up showing the Thank You page always, whenever the Decline button is clicked.

southp replied:
Thanks for the clarification :)

southp commented on client/my-sites/checkout/concierge-session-nudge/concierge-support-session/index.jsx:
```diff
@@ -0,0 +1,221 @@
+					<div className="concierge-support-session__column-doodle">
+						<img
+							className="concierge-support-session__doodle"
+							alt=""
```
Let's fill in this alt text in the name of accessibility :)

niranjan-uma-shankar replied:
Fixed in [df60aa9](https://github.com/Automattic/wp-calypso/pull/33996/commits/df60aa99d4aeb47033d61ff3a92d8aecae22d942)

southp commented on client/my-sites/checkout/concierge-session-nudge/concierge-support-session/index.jsx:
```diff
@@ -0,0 +1,221 @@
+										},
+									}
+								) }
+							</b>{' '}
```
I'm wondering what is the trailing `{' '}` for? 🤔

niranjan-uma-shankar replied:
This adds a space between the two sentences. Do you suggest doing it a different way?

<img width="681" alt="Screenshot 2019-07-01 at 4 57 22 PM" src="https://user-images.githubusercontent.com/1269602/60433121-a6ee4e00-9c21-11e9-91fa-c19fa1167bd3.png">

southp replied:
Ah, got it. Sorry for the ignorance.

niranjan-uma-shankar commented:
> I've noticed one regression though. When the quick start session nudge shows up after purchasing a plan, there is a heading belonging to the subsequent checkout form:

Thanks! Fixed in [662e8be](https://github.com/Automattic/wp-calypso/pull/33996/commits/662e8be025273981908ddebb6fb16421ac6ac61f).

> Another issue is that there is a weird hash-like string below the session product in the cart summary:

This is an issue been around for a while. I've raised a bug ticket on this #34375 . Are you okay if I take it up as a follow up, since the bug is not introduced by this PR?

niranjan-uma-shankar commented:
> AFAIK, add-support-session is still used by HEs to direct users to for purchasing sessions.
Do you think it would be worth redirecting the old route to the new one, so that we don't have that dead URL that probably still exists in the wild?

👍 

> I still see "Hold tight, your site is being upgraded." Maybe that got reverted/

Ah yes, I reverted it based on @victorespigares's comment to keep that unchanged. I've updated the PR description to reflect this.
